src/java/org/apache/cassandra/index/sai/utils/PrimaryKey.java

```diff
@@ -0,0 +1,184 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.ClusteringComparator;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * Representation of the primary key for a row consisting of the {@link DecoratedKey} and
+ * {@link Clustering} associated with a {@link org.apache.cassandra.db.rows.Row}.
+ */
+public class PrimaryKey implements Comparable<PrimaryKey>
+{
+    private final Token token;
+    private final DecoratedKey partitionKey;
+    private final Clustering<?> clustering;
+    private final ClusteringComparator clusteringComparator;
+
+    public PrimaryKey(Token token,
+                      DecoratedKey partitionKey,
+                      Clustering<?> clustering,
+                      ClusteringComparator clusteringComparator)
+    {
+        this.token = token;
+        this.partitionKey = partitionKey;
+        this.clustering = clustering;
+        this.clusteringComparator = clusteringComparator;
+    }
+
+    /**
+     * Returns a {@link PrimaryKeyFactory} for creating {@link PrimaryKey} instances.
+     *
+     * @param clusteringComparator the {@link ClusteringComparator} used by the
+     *                             {@link PrimaryKeyFactory} for clustering comparisons
+     * @return a {@link PrimaryKeyFactory} for {@link PrimaryKey} creation
+     */
+    public static PrimaryKeyFactory factory(ClusteringComparator clusteringComparator)
+    {
+        return new PrimaryKeyFactory(clusteringComparator);
+    }
+
+    /**
+     * Returns the {@link Token} associated with this primary key.
+     *
+     * @return the {@link Token}
+     */
+    public Token token()
+    {
+        return token;
+    }
+
+    /**
+     * Returns the {@link DecoratedKey} associated with this primary key.
+     *
+     * @return the {@link DecoratedKey}
+     */
+    public DecoratedKey partitionKey()
+    {
+        return partitionKey;
+    }
+
+    /**
+     * Returns the {@link Clustering} associated with this primary key
+     *
+     * @return the {@link Clustering}
+     */
+    public Clustering clustering()
```

Comment:
```suggestion
    public Clustering<?> clustering()
```
This also removes a compiler warning in `QueryController#makeFilter()`